File: src/tmp/Java/src/com/dbc/algorithm/Pointer/LongestContinuousSubarrayWithAbsoluteDiffLessThanOrEqualToLimit.java
package com.dbc.algorithm.Pointer;

import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;
import java.util.TreeMap;

public class LongestContinuousSubarrayWithAbsoluteDiffLessThanOrEqualToLimit {
    public static int longestSubarrayTreeMap(int[] nums, int limit) {
        int left = 0, right = 0, result = 0;
        TreeMap<Integer, Integer> map = new TreeMap<>();
        while (right < nums.length) {
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);
            if (map.lastKey() - map.firstKey() > limit) {
                if (map.get(nums[left]) > 1)
                    map.put(nums[left], map.getOrDefault(nums[left], 0) - 1);
                else
                    map.remove(nums[left]);
                left++;
            }
            result = Math.max(result, right - left + 1);
            right++;
        }
        return result;
    }

    public static int longestSubarrayQueue(int[] nums, int limit) {
        int left = 0, right = 0, result = 0;
        Deque<Integer> max = new LinkedList<>();
        Deque<Integer> min = new LinkedList<>();
        while (right < nums.length) {
            while (!max.isEmpty() && max.getLast() < nums[right]) {
                max.pollLast();
            }
            while (!min.isEmpty() && min.getLast() > nums[right]) {
                min.pollLast();
            }
            max.offerLast(nums[right]);
            min.offerLast(nums[right]);

            if (max.getFirst() - min.getFirst() > limit) {
                if (max.getFirst() == nums[left]) {
                    max.removeFirst();
                }
                if (min.getFirst() == nums[left]) {
                    min.removeFirst();
                }
                left++;
            }
            result = Math.max(result, right - left + 1);
            right++;
        }
        return result;
    }
}

File: core/tmp/Python/pointer/longest_continuous_subarray_with_absolute_diff_less_than_or_equal_to_limit.py
from collections import deque
from operator import le
from typing import List
from sortedcontainers import SortedList

def longestSubarrayTreeMap(nums: List[int], limit: int) -> int:
    left, right, result = 0, 0, 0
    remain = SortedList()
    while right < len(nums):
        remain.add(nums[right])
        if remain[-1] - remain[0] > limit:
            remain.remove(nums[left])
            left += 1
        result = max(result, right - left + 1)
        right += 1
    
    return result

def longestSubarrayQueue(nums: List[int], limit: int) -> int:
    length, left, right, result = len(nums), 0, 0, 0
    imax, imin = deque(), deque()
    while right < length:
        while imax and imax[-1] < nums[right]:
            imax.pop()
        while imin and imin[-1] >nums[right]:
            imin.pop()
        
        imax.append(nums[right])
        imin.append(nums[right])

        if imax[0] - imin[0] > limit:
            if nums[left] == imax[0]:
                imax.popleft()
            if nums[left] == imin[0]:
                imin.popleft()
            left += 1
        result = max(result, right - left + 1)
        right += 1
    return result

if __name__ == "__main__":
    print(longestSubarrayQueue(nums = [4,2,2,2,4,4,2,2], limit = 0))

